cs3client/file.py

```diff
@@ -334,3 +357,261 @@ def list_dir(
+            return None
+
+    def get_lock(self, auth_token: tuple, resource: Resource) -> cs3spr.Lock | dict | None:
+        """
+        Get the lock for the given filepath
+
+        :param auth_token: tuple in the form ('x-access-token', <token>) (see auth.get_token/auth.check_token)
+        :param resource: Resource to get lock from
+        :return: dictionary (KEYS: lock_id, type, app_name, user, expiration) (Success)
+        :return: None (No lock set)
+        :raises: NotFoundException (File not found)
+        :raises: AuthenticationException (Authentication Failed)
+        :raises: UnknownException (Unknown Error)
+        """
+        # fallback to xattr if the storage does not support locks
+        if self._config.lock_by_setting_attr and self._config.lock_not_impl:
+            return self._get_lock_using_xattr(auth_token, resource)
+
+        req = cs3sp.GetLockRequest(ref=resource.ref)
+        res = self._gateway.GetLock(request=req, metadata=[auth_token])
+
+        # if the storage does not support locks, get the lock using xattr
+        if res.status.code == cs3code.CODE_UNIMPLEMENTED and self._config.lock_by_setting_attr:
+            self._config.lock_not_impl = True
+            return self.get_lock(auth_token, resource)
+
+        self._status_code_handler.handle_errors(res.status, "get lock", resource.get_file_ref_str())
+        self._log.debug(f'msg="Invoked GetLock" {resource.get_file_ref_str()}" result="{res.status.trace}"')
+
+        # rebuild a dict corresponding to the internal JSON structure used by Reva
+        return {
+            "lock_id": res.lock.lock_id,
+            "type": res.lock.type,
+            "app_name": res.lock.app_name,
+            "user": (
+                {"opaque_id": res.lock.user.opaque_id, "idp": res.lock.user.idp, "type": res.lock.user.type}
+                if res.lock.user.opaque_id
+                else {}
+            ),
+            "expiration": {"seconds": res.lock.expiration.seconds},
+        }
+
+    def _refresh_lock_using_xattr(
+            self, auth_token: tuple, resource: Resource, app_name: str, lock_id: str | int, existing_lock_id: str | int = None
+    ) -> None:
+        """
+        Refresh the lock metadata for the given filepath
+
+        :param auth_token: tuple in the form ('x-access-token', <token>) (see auth.get_token/auth.check_token)
+        :param resource: Resource to refresh lock for
+        :param app_name: Application name
+        :param lock_id: metadata value to set
+        :param existing_lock_id: existing metadata vlue
+        :return: None (Success)
+        :raises: FileLockedException (File is locked)
+        :raises: NotFoundException (File not found)
+        :raises: AuthenticationException (Authentication Failed)
+        :raises: UnknownException (Unknown Error)
+        """
+        self._log.debug(f'msg="Using xattrs to execute RefreshLock" {resource.get_file_ref_str()}" value="{lock_id}"')
+        try:
```

Comment:
I do not understand why we're doing a try-catch here... If same app name and (not id or same id), you call set_attr and return. If not, raise exception.